File: apps/api-rs/src/rag/context.rs
use anyhow::Result;

use crate::models::rag::{ContextInput, EvidencePack};

#[async_trait::async_trait]
pub trait ContextAssembler: Send + Sync {
    async fn assemble(&self, input: ContextInput) -> Result<EvidencePack>;
}

pub struct SimpleContextAssembler;

impl SimpleContextAssembler {
    pub fn new() -> Self {
        Self
    }
}

impl Default for SimpleContextAssembler {
    fn default() -> Self {
        Self::new()
    }
}

#[async_trait::async_trait]
impl ContextAssembler for SimpleContextAssembler {
    async fn assemble(&self, input: ContextInput) -> Result<EvidencePack> {
        let mut lines = vec![];
        let mut selected = Vec::new();
        let mut used_chars = 0usize;
        for chunk in input.chunks {
            let chunk_chars = chunk.chunk.content.chars().count();
            if !selected.is_empty()
                && used_chars.saturating_add(chunk_chars) > input.max_context_chars.max(1)
            {
                continue;
            }
            used_chars = used_chars.saturating_add(chunk_chars);
            selected.push(chunk);
        }
        for (i, chunk) in selected.iter().enumerate() {
            let index = i + 1;
            let heading = if chunk.chunk.heading_path.is_empty() {
                String::new()
            } else {
                format!(" > {}", chunk.chunk.heading_path.join(" > "))
            };
            let page = if chunk.chunk.page_range.is_empty() {
                String::new()
            } else {
                format!(
                    "第{}",
                    chunk
                        .chunk
                        .page_range
                        .iter()
                        .map(|p| p.to_string())
                        .collect::<Vec<_>>()
                        .join("-")
                )
            };
            lines.push(format!(
                "[{index}] 文档: {title} {page}{heading}\n{content}",
                title = chunk.chunk.doc_title,
                content = chunk.chunk.content,
            ));
        }
        Ok(EvidencePack {
            chunks: selected,
            context_text: lines.join("\n\n"),
        })
    }
}

File: apps/api-bun/src/rag/context.ts
// 移植自 apps/api-rs/src/rag/context.rs —— SimpleContextAssembler：按 max_context_chars 预算拼装证据上下文
import type { ContextInput, EvidencePack } from '../models/rag.ts';
import type { ContextAssembler } from './types.ts';

export class SimpleContextAssembler implements ContextAssembler {
  constructor() {}

  componentName(): string {
    return 'simple-context-assembler';
  }

  async assemble(input: ContextInput): Promise<EvidencePack> {
    const lines: string[] = [];
    const selected = [] as EvidencePack['chunks'];
    let usedChars = 0;
    for (const chunk of input.chunks) {
      const chunkChars = Array.from(chunk.chunk.content).length;
      if (selected.length > 0 && usedChars + chunkChars > Math.max(input.max_context_chars, 1)) {
        continue;
      }
      usedChars += chunkChars;
      selected.push(chunk);
    }
    for (const [i, chunk] of selected.entries()) {
      const index = i + 1;
      const heading =
        chunk.chunk.heading_path.length === 0
          ? ''
          : ` > ${chunk.chunk.heading_path.join(' > ')}`;
      const page =
        chunk.chunk.page_range.length === 0
          ? ''
          : `第${chunk.chunk.page_range.map((p) => p.toString()).join('-')}`;
      lines.push(
        `[${index}] 文档: ${chunk.chunk.doc_title} ${page}${heading}
${chunk.chunk.content}`,
      );
    }
    return { chunks: selected, context_text: lines.join('\n\n') };
  }
}
